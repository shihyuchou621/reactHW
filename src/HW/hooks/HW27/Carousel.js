import cx from 'classnames';
import React, { useState, useEffect } from 'react';

import style from './Carousel.module.css';

export default function Carousel(props) {

  const [isMove, setIsMove] = useState(false);
  const [whichShow, setWhichShow] = useState(0);
  const [itemList, setItemList] = useState(props.children);

  let timeout;

  const cycle = () => {
    const { time } = props;
    timeout = setInterval(() => {
      setIsMove(true);
      setWhichShow(whichShow => ( whichShow + 1 ) % itemList.length);

      setTimeout(() => {
        setItemList(itemList => [
          ...itemList.slice(1),
          ...itemList.slice(0,1)
        ]);
        setIsMove(false);
      }, time / 3);
    }, time);
  };

  useEffect(() => {
    cycle();
    return () => {
      clearTimeout(timeout);
    };
  }, []);

  return (
    <div className={style.wrapper}>
      {itemList.map((item, index) =>
        <div key={index} className={cx(style.item, {
          [style.animationHide]: index === 0 && isMove,
          [style.animationShow]: index === 1 && isMove})}>
          {item}
        </div>
      )}
      {props.showDots &&
          <div className={style.dotsWrapper}>
            {[...itemList.keys()].map(index =>
              <div key={index} className={cx(style.dots, {
                [style.dotShow]: index === whichShow,
              })} />
            )}
          </div>
      }
    </div>
  );
}
