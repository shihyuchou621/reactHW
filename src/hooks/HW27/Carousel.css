.wrapper {
  width: 500px;
  height: 300px;
  position: relative;
  display: inline-block;
  background-color: rgb(200, 200, 200);
  overflow: hidden;
}

@keyframes hide {
  from { left: 0%;}
  to { left: -100%;}
}

@keyframes show {
  from { left: 100%;}
  to { left: 0%;}
}

.item {
  width: 100%;
  height: 100%;
  background-color: rosybrown;
  text-align: center;
  /* display: flex; */
}

/* .item:nth-of-type(2) {
  background-color: #ff0;
} */

.animationShow {
  left: 0;
  position: absolute;
  animation: show 500ms;
  animation-fill-mode: forwards;
  /* animation-iteration-count: infinite; */
}

.animationHide {
  left: 0;
  position: absolute;
  animation: hide 500ms;
  animation-fill-mode: forwards;
  /* animation-iteration-count: infinite; */
}

.dotsWrapper {
  width: 100%;
  bottom: 20px;
  display: flex;
  justify-content: center;
  position: absolute;
}

.dots {
  width: 10px;
  height: 10px;
  margin: 5px;
  border-radius: 100%;
  background-color: gray;
  display: inline-block;
}

.dotShow {
  background-color: white;
}