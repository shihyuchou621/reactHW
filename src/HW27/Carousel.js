import cx from 'classnames';
import React, { Component } from 'react'

import './Carousel.css';

export default class Carousel extends Component {
  //constructor要使用props一定要用下面這種寫法
  constructor(props) {
    super();
    this.state = {
      whichShow: 0,
      isMove: false,
      itemList: props.children,
    }
  }

  cycle = () => {
    const { time } = this.props;
    const { itemList, whichShow } = this.state;

    setInterval(() => {

      this.setState({
        isMove: true,
        whichShow: ( whichShow + 1 ) % itemList.length,
      })

      setTimeout(() => {
        this.setState({
          itemList: [
            ...itemList.slice(1),
            ...itemList.slice(0,1)
          ],
          isMove: false,
        })
      }, time / 3);
    }, time);
    
  }

  componentDidMount() {
    this.cycle();
  }
  
  render() {
    const {itemList, isMove, whichShow} = this.state;
    return (
      <div className="wrapper">
        {itemList.map((item, index) =>
          <div key={index} className={cx("item ", {
            animationHide: index === 0 && isMove,
            animationShow: index === 1 && isMove})}>
            {item}
          </div>
        )}
        {this.props.showDots &&
          <div className="dotsWrapper">
            {itemList.map((item, index) => 
              <div key={index} className={cx("dots", {
                dotShow: index === whichShow,
              })} />
            )}
          </div>
        }
      </div>
    )
  }
}
