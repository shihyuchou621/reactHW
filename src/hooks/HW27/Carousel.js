import cx from 'classnames';
import React, { useState, useEffect } from 'react';

import './Carousel.css';

export default function Carousel(props) {

  const [isMove, setIsMove] = useState(false);
  const [whichShow, setWhichShow] = useState(0);
  const [itemList, setItemList] = useState(props.children);

  let timeout;

  const cycle = () => {
    const { time } = props;
    timeout = setInterval(() => {
      setIsMove(true);
      setWhichShow(whichShow => ( whichShow + 1 ) % itemList.length);

      setTimeout(() => {
        setItemList(itemList => [
          ...itemList.slice(1),
          ...itemList.slice(0,1)
        ]);
        setIsMove(false);
      }, time / 3);
    }, time);
  };

  useEffect(() => {
    cycle();
    return () => {
      clearTimeout(timeout);
    };
  }, []);

  return (
    <div className="wrapper">
      {itemList.map((item, index) =>
        <div key={index} className={cx("item ", {
          animationHide: index === 0 && isMove,
          animationShow: index === 1 && isMove})}>
          {item}
        </div>
      )}
      {props.showDots &&
          <div className="dotsWrapper">
            {[...itemList.keys()].map(index =>
              <div key={index} className={cx("dots", {
                dotShow: index === whichShow,
              })} />
            )}
          </div>
      }
    </div>
  );
}
