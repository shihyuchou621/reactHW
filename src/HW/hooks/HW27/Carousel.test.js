import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Carousel from './Carousel';

afterEach(cleanup);

jest.useFakeTimers();

test('animation should work', () => {
  act(() => {
    const { container } = render(
      <Carousel
        time={1500}
        showDots={true}
      >
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Carousel>
    );
    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();

    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();

    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();
  });
});


test('showDots is false', () => {
  act(() => {
    const { container } = render(
      <Carousel
        time={1500}
        showDots={false}
      >
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Carousel>
    );
    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();

    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();

    jest.advanceTimersByTime(1550);
    expect(container).toMatchSnapshot();
  });
});
