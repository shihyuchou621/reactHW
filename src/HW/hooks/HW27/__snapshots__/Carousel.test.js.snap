// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animation should work 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item animationHide"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="item animationShow"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="dotsWrapper"
    >
      <div
        class="dots"
      />
      <div
        class="dots dotShow"
      />
      <div
        class="dots"
      />
    </div>
  </div>
</div>
`;

exports[`animation should work 2`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="dotsWrapper"
    >
      <div
        class="dots"
      />
      <div
        class="dots dotShow"
      />
      <div
        class="dots"
      />
    </div>
  </div>
</div>
`;

exports[`animation should work 3`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="dotsWrapper"
    >
      <div
        class="dots"
      />
      <div
        class="dots"
      />
      <div
        class="dots dotShow"
      />
    </div>
  </div>
</div>
`;

exports[`showDots is false 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item animationHide"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="item animationShow"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
  </div>
</div>
`;

exports[`showDots is false 2`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        1
      </div>
    </div>
  </div>
</div>
`;

exports[`showDots is false 3`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="item"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="item"
    >
      <div>
        2
      </div>
    </div>
  </div>
</div>
`;
